### ISSUE TITLE
Watchdog caught collective operation timeout when using Multi-GPUs

### ISSUE THREAD

**Issue by behroozazarkhalili** [ultralytics/ultralytics#1439]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.


### YOLOv8 Component

Training, Detection, Multi-GPU

### Bug

I got the following error when using two RTX 4090 GPUS:                                                                                                                                                                                                        
```
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
[E ProcessGroupNCCL.cpp:821] [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=1, OpType=BROADCAST, Timeout(ms)=1800000) ran for 1807146 milliseconds before timing out.                                                 
[E ProcessGroupNCCL.cpp:821] [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=2, OpType=ALLGATHER, Timeout(ms)=1800000) ran for 1807168 milliseconds before timing out. 
```
```
[E ProcessGroupNCCL.cpp:456] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
[E ProcessGroupNCCL.cpp:461] To avoid data inconsistency, we are taking the entire process down.
```

The issue happened when we had the AMP check. Although using only one GPU everything seems fine, using two GPUs raised the above error. 

Moreover, when setting GPUS with the `device` variable in `model.train`, I cannot see the progress bar for epochs.

### Environment

```
Ultralytics YOLOv8.0.53 🚀 
Python-3.10.9 
torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB) CUDA:1 (NVIDIA GeForce RTX 4090, 24215MiB)
```

### Minimal Reproducible Example

```python
from ultralytics import YOLO
model = YOLO("yolov8s.yaml")

# Train the model
model.train(data="coco128.yaml", epochs=50, imgsz=640, device="0, 1")
```

### Additional

The problem should be related to the following issues:
[Multi-gpu training timeout error work nccl](https://discuss.pytorch.org/t/multi-gpu-training-timeout-error-worknccl-optype-allgather-timeout/169435)
[Watchdog caught collective operation timeout](https://github.com/WongKinYiu/yolov7/issues/714)


### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@behroozazarkhalili we're training DDP models currently with 8.0.54 without any issues, with AMP checks passing. If you're having problems with your local environment you could try disabling the AMP check completely (i.e. just set self.amp = True) or use our Docker image below.

## Environments

YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

**Comment by glenn-jocher** [comment]
@AyushExel seems like based on above AMP could use an argument to force it's behavior, i.e. amp=None default and can be set to amp=True/False to force one or the other without any checks?

**Comment by mehran66** [comment]
I have had the same issue and did figure out that there is an issue with my data that leads to this erorr. Have not found out what is the issue with my data but have been able to run the training by the following change: 

        if world_size > 1:
            self.model = DDP(self.model, device_ids=[RANK], **find_unused_parameters=True**)
            
https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/engine/trainer.py#LL218:~:text=)-,if,-world_size

**Comment by Anita-Hailey** [comment]
@behroozazarkhalili How to solve this problem?

**Comment by AlbertoOddness** [comment]
Currently I am facing the exact same issue. Any solutions?

**Comment by glenn-jocher** [comment]
@AlbertoOddness, we have a few suggestions on how to solve this issue:

1. Upgrade PyTorch to the latest stable release version. 
2. If upgrading PyTorch doesn't work, you could also try switching to the "nccl" backend. This can be done by setting the environment variable `NCCL_SOCKET_IFNAME` to the network interface you want to use for communication before launching your script.
3. Another option worth trying is to use a different batch size, which can sometimes help with timeout issues.
4. If none of the above works, you may want to try running your code on a different machine with different hardware to rule out any possible hardware issues.

I hope this helps! Let me know if you have any further questions.

**Comment by mehran66** [comment]
@AlbertoOddness @glenn-jocher I had the same issue and spent a lot of time explored many potential solutions including checking and fixing data issue, upgrading the pytorch and cuda to the latest version, trying new AWS EC2 instances, changing the parameters in the config file, etc and finally figured out that the issue is related to how I install ultralytics:

I was installing ultralytics in a folder like the following so I can easily test changes in the code if needed:

  git clone https://github.com/ultralytics/ultralytics
  cd ultralytics
  pip install --user -e '.[dev]'

but after installing it like `pip install ultralytics`, my multigpu issue got resolved.

**Comment by glenn-jocher** [comment]
@mehran66, thank you for sharing your experience with regards to fixing the multi-GPU training issue. It's great to hear that installing Ultralytics via pip install resolved the problem. It's important to note that local installation changes and modification made to the project require the appropriate dependencies to be installed, which may cause dataset incompatibility and lead to the issue you experienced. If you encounter any further issues, feel free to post it and we'll be happy to help.

**Comment by glenn-jocher** [comment]
Removing TODO as I'm not able to reproduce and a fix is mentioned in https://github.com/ultralytics/ultralytics/issues/1439#issuecomment-1583010583

**Comment by glenn-jocher** [comment]
@mehran66 @behroozazarkhalili @Anita-Hailey @AlbertoOddness see fix for this problem in https://github.com/ultralytics/ultralytics/issues/1439#issuecomment-1583010583, and also I'd always highly recommend everyone train in our Docker image for max compatibility. This is where we do all our DDP trainings :)

## Environments

YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

**Comment by rumjie** [comment]
I solved the problem. 

in the package that you installed, change `def _setup_ddp' function in `ultralytics/yolo/engine/trainer.py` like below:
'os.environ['NCCL_BLOCKING_WAIT'] = '0'  # not to enforce timeout
 dist.init_process_group('nccl' if dist.is_nccl_available() else 'gloo',
                                timeout=timedelta(seconds=7200000), # 1800000
                                rank=RANK,
                                world_size=world_size)'

**Comment by rumjie** [comment]
> @rumjie great job on solving the problem!
> 
> It looks like the issue was related to the timeout during the initialization of the distributed training environment. By modifying the `_setup_ddp` function in `ultralytics/yolo/engine/trainer.py`, you set `NCCL_BLOCKING_WAIT` to `0` to disable the timeout enforcement, and increased the timeout limit to a larger value (7200000 seconds in your case).
> 
> Additionally, you mentioned that the dataset you used was quite large (2.2 TB) and took a long time to scan, which might have caused the timeout issue.
> 
> I'm glad to hear that the solution worked for you. Hopefully, this information can help others who face a similar problem. If you have any further questions or issues, feel free to ask.
Thanks a lot for comment! 

I got minor question about this - is there any specific reason for limiting the time during the process? And is this limit only applied on multi-GPU training? (One of my trial was using 1 GPU, and I had no idea why scanning was done with this) 

I guess people who handles custom dataset might have similar issue with default time limit - and I requested PR.

**Comment by glenn-jocher** [comment]
Thank you for sharing your solution, @rumjie!

Regarding your question about the timeout limit during the initialization of the distributed training environment, the reason behind setting a timeout is to ensure that the initialization process completes within a reasonable time frame. By default, there is a timeout of 30 minutes (1800 seconds) enforced. However, in your case, the dataset you used was quite large (2.2 TB), which resulted in a longer scanning time. This led to the timeout issue during the initialization.

The timeout limit is not specific to multi-GPU training but applies to all distributed training scenarios. When using a single GPU, you might not have encountered the scanning issue due to the shorter initialization time. However, changing the timeout and disabling the enforcement can be beneficial for cases where datasets take a longer time to scan, as you mentioned.

Your request for a PR regarding the timeout limit for custom datasets is valid as it can help others facing similar issues. This modification would provide more flexibility in handling datasets of varying sizes and processing times. 

If you have any further questions or need assistance, please don't hesitate to ask.

**Comment by glenn-jocher** [comment]
@behroozazarkhalili @rumjie great news 😃! Your original issue may now be resolved ✅ in PR #3343. This PR increases the NCCL timeout from 1 to 3 hours.

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by thelongestusernameofall** [comment]
solution in https://github.com/NVIDIA/nccl/issues/426  works.

export NCCL_IB_GID_INDEX=3    solved my problem.

**Comment by glenn-jocher** [comment]
Fantastic, @thelongestusernameofall! I'm glad to hear you've found a solution that works for you, and I appreciate you sharing it with the community. Setting the `NCCL_IB_GID_INDEX` environment variable can indeed help when experiencing network communication issues during distributed training with NCCL. 

Keep in mind that the optimal `NCCL_IB_GID_INDEX` value can vary depending on your system configuration and network setup. It's excellent that in your case, setting it to 3 resolved the issue.

Should you come across any further issues or have new insights you'd like to share, please feel free to reach out. Your contributions are valuable in improving the experience for everyone using Ultralytics YOLOv8. Happy coding! 🚀

**Comment by JerryDaHeLian** [comment]
> @AlbertoOddness, we have a few suggestions on how to solve this issue:
> 
> 1. Upgrade PyTorch to the latest stable release version.
> 2. If upgrading PyTorch doesn't work, you could also try switching to the "nccl" backend. This can be done by setting the environment variable `NCCL_SOCKET_IFNAME` to the network interface you want to use for communication before launching your script.
> 3. Another option worth trying is to use a different batch size, which can sometimes help with timeout issues.
> 4. If none of the above works, you may want to try running your code on a different machine with different hardware to rule out any possible hardware issues.
> 
> I hope this helps! Let me know if you have any further questions.

当我在各个rank 上使用不同的batch_size 时，会报Watchdog caught collective operation timeout  错误，是不是必须在所有的rank 上都要设置相同的batch_size？

**Comment by cucdengjunli** [comment]
good question

**Comment by glenn-jocher** [comment]
@cucdengjunli yes, for distributed training using YOLOv8 and NCCL, it's essential that all ranks use the same `batch_size` to ensure proper synchronization and data distribution across the GPUs. Different batch sizes can lead to inconsistencies and errors like the one you're experiencing. Here's a quick example on setting `batch_size`:

```python
model.train(data='coco128.yaml', epochs=100, imgsz=640, batch_size=16)
```

Make sure to adjust `batch_size` according to your system's capabilities to maximize performance. Happy training! 🚀

**Comment by kzos** [comment]
> I solved the problem.
> 
> in the package that you installed, change `def _setup_ddp` function in `ultralytics/yolo/engine/trainer.py` like below:
> 
> ```
> os.environ['NCCL_BLOCKING_WAIT'] = '0'  # not to enforce timeout
> dist.init_process_group('nccl' if dist.is_nccl_available() else 'gloo',
>                                 timeout=timedelta(seconds=7200000), # was 1800000
>                                 rank=RANK,
>                                 world_size=world_size)
> ```
> 
> In my case the dataset I used was too large (2.2 TB) and time for scanning was too long. Hope you this works

This fix worked , thanks again,

**Comment by pderrenger** [comment]
Great to hear that the fix worked for you! If you encounter any other issues or have further questions, feel free to reach out. Always ensure you're using the latest version of the package to benefit from the latest fixes and improvements. Happy training!

**Comment by ZhonggaiWang** [comment]
> I solved the problem.
> 
> in the package that you installed, change `def _setup_ddp` function in `ultralytics/yolo/engine/trainer.py` like below:
> 
> ```
> os.environ['NCCL_BLOCKING_WAIT'] = '0'  # not to enforce timeout
> dist.init_process_group('nccl' if dist.is_nccl_available() else 'gloo',
>                                 timeout=timedelta(seconds=7200000), # was 1800000
>                                 rank=RANK,
>                                 world_size=world_size)
> ```
> 
> In my case the dataset I used was too large (2.2 TB) and time for scanning was too long. Hope you this works

it helps , thank u

**Comment by pderrenger** [comment]
@ZhonggaiWang i'm glad to hear that the solution worked for you. If you encounter any further issues or have additional questions, please ensure you are using the latest version of the Ultralytics package. This helps to benefit from the latest fixes and improvements. Thank you for your feedback and happy training!

**Comment by glenn-jocher** [comment]
I'm glad to hear the solution worked for you! If you have any more questions or need further assistance, feel free to ask.

**Comment by glenn-jocher** [comment]
@kzos great to hear that the solution worked for you! If you have any more questions or need further assistance, feel free to ask.

**Comment by glenn-jocher** [comment]
@kzos great to hear that the solution worked for you! If you have any more questions or need further assistance, feel free to ask.

**Comment by glenn-jocher** [comment]
I'm glad to hear the solution worked for you! If you have any more questions or need further assistance, feel free to ask.
